python/packages/durabletask/DESIGN.md

```diff
@@ -0,0 +1,249 @@
+
+### 2. State Management (`_durable_agent_state.py`)
+
+*   **Goal**: Maintain 100% schema compatibility with `agent-framework-azurefunctions`.
+*   **Implementation**: Direct port of `packages/azurefunctions/agent_framework_azurefunctions/_durable_agent_state.py`.
+
+### 3. The Agent Entity (`_entities.py`)
+
+We will implement a class `AgentEntity` that inherits from `durabletask.entities.DurableEntity`.
+
+**Important**: This will be ported from `azurefunctions` package too but with slight modifications, details TBD.
+
+### 4. The Worker Wrapper (`_worker.py`)
+
+The `DurableAIAgentWorker` wraps an existing `durabletask` worker instance.
+
+```python
+class DurableAIAgentWorker:
+    def __init__(self, worker: TaskHubGrpcWorker):
+        self._worker = worker
+        self._registered_agents: dict[str, AgentProtocol] = {}
+
+    def add_agent(self, agent: AgentProtocol) -> None:
+        """Registers an agent with the worker.
+        
+        Uses the factory pattern to create an AgentEntity class with the agent
+        instance injected, then registers it with the durabletask worker.
+        """
+        # Store the agent reference
+        self._registered_agents[agent.name] = agent
+        
+        # Create a configured entity class using the factory
+        entity_class = create_agent_entity(agent)
+        
+        # Register the entity class with the worker
+        # The worker.add_entity method takes a class or function
+        self._worker.add_entity(entity_class)
+
+    def start(self):
+        """Start the worker to begin processing tasks."""
+        self._worker.start()
+
+    def stop(self):
+        """Stop the worker gracefully."""
+        self._worker.stop()
+```
+
+### 5. The Mixin (`_utils.py`)
+
+```python
+class GetDurableAgentMixin:
+    """Mixin to provide get_agent interface."""
+    
+    def get_agent(self, agent_name: str) -> 'DurableAIAgent':
+        raise NotImplementedError
+```
+
+### 6. The Client Wrapper (`_client.py`)
+
+The `DurableAIAgentClient` is for external clients (e.g., FastAPI, CLI).
+
+```python
+class DurableAIAgentClient(GetDurableAgentMixin):
+    def __init__(self, client: TaskHubGrpcClient):
+        self._client = client
+
+    async def get_agent(self, agent_name: str) -> 'DurableAIAgent':
+        """Retrieves a DurableAIAgent shim.
+        
+        Validates existence by attempting to fetch entity state/metadata.
+        """
+        # Validation logic using self._client.get_entity(...)
+        # ...
+        return DurableAIAgent(self, agent_name)
+
+    def run_agent(self, agent_name: str, message: str, **kwargs) -> 'Task':
+        """Runs agent via signal + poll and returns a Task wrapper."""
+        # Returns a ClientTask (wrapper around asyncio.Task)
+        pass
+```
+
+### 7. The Orchestrator Wrapper (`_orchestrator.py`)
+
+The `DurableAIAgentOrchestrator` is for use *inside* orchestrations.
+
+```python
+class DurableAIAgentOrchestrator(GetDurableAgentMixin):
```

Comment:
This isn't an "orchestrator" - an orchestrator is a function. This is an _orchestration context_, so we should name it accordingly - e.g., `DurableAIAgentOrchestrationContext`.

Reply:
renamed to `DurableAIAgentOrchestrationContext`